iroh-dns-server/src/dns/node_authority.rs

```diff
@@ -0,0 +1,291 @@
+use std::{
+    collections::{btree_map, BTreeMap},
+    fmt,
+    sync::Arc,
+};
+
+use anyhow::{bail, Result};
+use async_trait::async_trait;
+use hickory_proto::{
+    op::ResponseCode,
+    rr::{LowerName, Name, RecordSet, RecordType, RrKey},
+};
+use hickory_server::{
+    authority::{
+        AuthLookup, Authority, LookupError, LookupOptions, LookupRecords, MessageRequest,
+        UpdateResult, ZoneType,
+    },
+    server::RequestInfo,
+    store::in_memory::InMemoryAuthority,
+};
+
+use iroh_metrics::inc;
+use parking_lot::RwLock;
+use pkarr::SignedPacket;
+use tracing::{debug, trace};
+
+use crate::util::{record_set_append_origin, signed_packet_to_hickory_records_without_origin};
+use crate::{metrics::Metrics, store::SignedPacketStore};
+
+pub enum PacketSource {
+    PkarrPublish,
+    Mainline,
+}
+
+pub type PublicKeyBytes = [u8; 32];
+
+pub struct NodeAuthority {
+    serial: u32,
+    primary_origin: LowerName,
+    all_origins: Vec<Name>,
+
+    store: SignedPacketStore,
+    static_authority: InMemoryAuthority,
+    zones: RwLock<BTreeMap<PublicKeyBytes, PkarrZone>>,
+}
+
+struct PkarrZone {
+    timestamp: u64,
+    records: BTreeMap<RrKey, Arc<RecordSet>>,
+}
+impl PkarrZone {
+    fn from_signed_packet(signed_packet: &SignedPacket) -> Result<Self> {
+        let (_label, records) =
+            signed_packet_to_hickory_records_without_origin(signed_packet, |_| true)?;
+        Ok(Self {
+            records,
+            timestamp: *signed_packet.timestamp(),
+        })
+    }
+
+    fn older_than(&self, signed_packet: &SignedPacket) -> bool {
+        *signed_packet.timestamp() > self.timestamp
+    }
+
+    fn records(&self) -> &BTreeMap<RrKey, Arc<RecordSet>> {
+        &self.records
+    }
+}
+
+impl NodeAuthority {
+    pub fn new(
+        store: SignedPacketStore,
+        static_authority: InMemoryAuthority,
+        primary_origin: Name,
+        additional_origins: Vec<Name>,
+        serial: u32,
+    ) -> Result<Self> {
+        let origins = {
+            let mut o = Vec::with_capacity(additional_origins.len());
+            o.push(primary_origin.clone());
+            o.extend_from_slice(&additional_origins);
+            o
+        };
+        let this = Self {
+            static_authority,
+            primary_origin: primary_origin.into(),
+            all_origins: origins,
+            serial,
+            store,
+            zones: Default::default(),
+        };
+        for packet in this.store.iter()? {
```

Comment:
This means that you have to iterate over every packet in the store on startup?

1. If the "store" is the DHT, you don't have a way to iterate over all records.

2. Imagine somebody is spamming you with records. The few bytes in the store are not going to be a problem, if inserting is fast (see #3 ). But if you have to do this for each startup it might be a problem.

I don't really know how hickory dns works, but can we do this on demand if we get an actual request?

Reply:
Having looked at this some more, this seems to be just an optimisation that should be removed or at least made lazy. We should not assume that we have a way to iterate over all pkarr packets, otherwise the idea to use the DHT goes out of the window. Instead, either don't do this optimization at all, or do it lazily on request for a key.

Reply:
This loads all records in the store, yes. The DNS server will then operate in-memory only for lookups, but will be limited by RAM size on the amount of records possible. The record size will not be more than 1000 bytes per signed packet because of pkarr limit, so with 1 million records it'd need about 1GB of memory.

That would have to be weighed against redb lookups for each DNS query, plus a few allocations per lookup likely for constructing the hickory data structures.

So I think it makes sense to keep it mostly in-memory, but it would be better to organize this as a LRU cache which falls back to redb, not as "load all on start only". Then the cache size does not determine the maximum amount of records stored.

Reply:
When we involve mainline, a tiered resolver makes sense I think that resolves until something found, first in-memory cache, then redb, then mainline. I will think some more how to structure this in the code.

Reply:
This is fixed in #9